### PR TITLE
[RFC] PipeWire audio driver

### PR DESCRIPTION
**Pull request by sykhro**
An implementation for an audio driver using the (relatively) new PipeWire libraries.
So far it has only been tested with playback, I'm planning on setting up a benchmark for latency soon.

As of now the functionality is basic but it can easily be extended to support more channels and audio formats.

### CODE REVIEW

**Comment by derselbst** [comment]
Oh, very nice! @SubhadeepJasu You're welcome to test this and give feedback.

**Comment by SubhadeepJasu** [comment]
Audio is a bit glitchy when using higher number of channels. What settings would you recommend for this driver for average of 8 channels max 16 channel playback?

**Comment by SubhadeepJasu** [comment]
These are my current settings:

```C
fluid_settings_setstr(style_synth_settings, "audio.driver", "pipewire");
fluid_settings_setint(style_synth_settings, "audio.period-size", 256);
fluid_settings_setint(style_synth_settings, "audio.realtime-prio", 1);
fluid_settings_setnum(style_synth_settings, "synth.gain", 2);
fluid_settings_setnum(style_synth_settings, "synth.overflow.percussion", 5000.0);
fluid_settings_setstr(style_synth_settings, "synth.midi-bank-select", "gs");
fluid_settings_setstr(style_synth_settings, "audio.pipewire.media-role", "Production");
fluid_settings_setstr(style_synth_settings, "audio.pipewire.media-type", "Audio");
```

**Comment by SubhadeepJasu** [comment]
Update: It looks like lowering the period size reduces the glitches. But there's still a lot of crackling audio. Not sure how to fix that.

**Comment by sykhro** [comment]
> Update: It looks like lowering the period size reduces the glitches. But there's still a lot of crackling audio. Not sure how to fix that.

Sounds like xruns, do you get the same with pulseaudio? Does setting `audio.realtime-prio` to 1 improve the situation? Does setting "Pro Audio" as pipewire profile change anything? (it can be done via pavucontrol).
I'll try and do some profiling to see if there's something I'm missing here.

**Comment by SubhadeepJasu** [comment]
Setting "Pro Audio" in pavucontrol definitely fixes the issue, or at least I don't hear any crackling at all. I need to test it more to confirm. The priority settings only seem to affect the latency (which improves playing the instruments, priority 1 causes the synth output to be instant with no noticable delay, but without priority = 1 there is some delay), but I found no effect other than that when the style engine is playing with like 7 or 8 channels active and 16 channels in total.

As for the previous question. Pulseaudio had the opposite effect. Increasing period-size caused the audio crackling to go away, but if I increase it too much, there is much noticable latency which makes it difficult to play the app (as in play instruments in the app with a midi controller).

**Comment by sykhro** [comment]
> The priority settings only seem to affect the latency

yes, that is intended.
I will try to ask around as the PipeWire API is severely undocumented but using the "Pro Audio" profile is probably the intended way when dealing with low-latency applications like yours.

**Comment by SubhadeepJasu** [comment]
I need to know a bit more about "Pro Audio" profile. Is it something that is available to every distro regardless of whether pipewire is installed or not? Is it okay to recommend end users to configure pavucontrol to use this profile or would it be best that find a solution that works out of box with little configuration from end user?

**Comment by sykhro** [comment]
> I need to know a bit more about "Pro Audio" profile. Is it something that is available to every distro regardless of whether pipewire is installed or not? Is it okay to recommend end users to configure pavucontrol to use this profile or would it be best that find a solution that works out of box with little configuration from end user?

It was added with PipeWire 0.3.22 (released 6 months ago), it's used to "have raw device access with the
maximum number of channels and no mixer controls"; it is, I believe, the option for people doing any sort of realtime audio work as it achieves very low latencies without straining the cpu much. I will try to see if there's a way to do this programmatically so that users don't have to do it, although I believe it's not supported. It is left, most likely, to people that know what they need

**Comment by SubhadeepJasu** [comment]
I see. If we don't find a way to use pro audio out of box, I'll somehow have to direct users to pavucontrol. I plan to provide a simple button to launch pavucontrol. (if you know some other ways to do the same pls let me know because pavucontrol doesn't seem to be installed out of box in some distros). Anyway you are doing great work. Let's focus on the driver and performance for now.

**Comment by mawe42** [comment]
@sykhro Thanks for this PR! I haven't looked at the code yet, but something you wrote got my attention. You said:
> [...] as the PipeWire API is severely undocumented [...]

That made me have a quick look at PipeWire with regard to API stability and I found [this](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#is-pipewire-ready-yet):
> The API/ABI has been declared stable since version 0.3.

That does sound ok. However, the FAQ also contains the [following](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#what-audio-api-do-you-recommend-to-use):
> ### What Audio API Do You Recommend To Use?
> The situation is a bit like GUI toolkits. There are many, each with different use cases. Nobody uses the native display server protocols directly (X11, Wayland) but always through an abstraction layer (GTK, Qt, etc).
> **We recommend that you continue to use PulseAudio, JACK and ALSA API's for now.**

So I must admit I do wonder if it is a good idea to implement a PipeWire driver in FluidSynth after all, especially if the PipeWire devs themselves recommend against using the native API.

**Comment by sykhro** [comment]
> So I must admit I do wonder if it is a good idea to implement a PipeWire driver in FluidSynth after all, especially if the PipeWire devs themselves recommend against using the native API.

I guess it could be useful on embedded systems where you don't want to mess around too much with getting JACK to work(?). Besides that, honestly, I'm not too sure either. This was more of an exercise to see what the API looks like and see if there's any benefits wrt to latency (compared to pulseaudio, at least)

**Comment by mawe42** [comment]
Ah, I just noticed that @SubhadeepJasu [found the same info](https://github.com/FluidSynth/fluidsynth/discussions/976#discussioncomment-1299280) 8 days ago already :-) 

> I guess it could be useful on embedded systems where you don't want to mess around too much with getting JACK to work(?).

Well, considering that the PipeWire devs say that PipeWire will "possibly" be as good as JACK (especially with regard to latency), JACK (or straight ALSA) might be the better choice for embedded until PipeWire is more mature. I don't know...

> Besides that, honestly, I'm not too sure either. This was more of an exercise to see what the API looks like and see if there's any benefits wrt to latency (compared to pulseaudio, at least)

Ah, I see. Ok.... maybe keep this driver around, either in this PR, in a pipewire-devel branch or in a forked repo or yours? What do you think, @derselbst ?

**Comment by SubhadeepJasu** [comment]
I see minor improvements when using pipewire core as opposed to pipewire pulse. Like the fluidplayer plays at the correct tempo now (previously it was a slightly slower) . With the pro audio profile it seems like the best I have heard. Also have low level access to the core api which is also very well managed it nice. And it's definitely waay easy than jack. And I can hear other applications while doing production work.

**Comment by derselbst** [comment]
> maybe keep this driver around, either in this PR, in a pipewire-devel branch or in a forked repo or yours? What do you think, @derselbst ?

I would leave this decision to @sykhro and @SubhadeepJasu. They seem to have a use-case for it. And once they have completed their testing and find it useful, I'm willing to merge this into master. We will probably (have to) keep it around in this PR for a few more weeks - I also want to give it a test, but I'm not sure when I find time for this.

Additionally, we should ensure that this driver will be built by our Linux CI build. So if the API unexpectedly changes, we see it instantly.

If everything goes well and we do merge it, we could also think about deprecating the OSS driver in turn.

**Comment by SubhadeepJasu** [comment]
Crackling is back even with Pro Audio profile selected. Not sure why tho. And the period size is set to the lowest value, anything higher causes more crackling. May be something to do with that.

**Comment by SubhadeepJasu** [comment]
> Crackling is back even with Pro Audio profile selected. Not sure why tho. And the period size is set to the lowest value, anything higher causes more crackling. May be something to do with that. May be my computer is weird. Or I need to recompile fluidsynth again.

Just Tested on my new laptop. Now it sounds great. No sound issues. So that probably was my old laptop dying.

**Comment by sykhro** [comment]
> > maybe keep this driver around, either in this PR, in a pipewire-devel branch or in a forked repo or yours? What do you think, @derselbst ?
> 
> I would leave this decision to @sykhro and @SubhadeepJasu. They seem to have a use-case for it. And once they have completed their testing and find it useful, I'm willing to merge this into master. We will probably (have to) keep it around in this PR for a few more weeks - I also want to give it a test, but I'm not sure when I find time for this.
> 
> Additionally, we should ensure that this driver will be built by our Linux CI build. So if the API unexpectedly changes, we see it instantly.
> 
> If everything goes well and we do merge it, we could also think about deprecating the OSS driver in turn.

I'll try to do some benchmarking asap. For me it's a nice-to-have but definitely not super high priority

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:196]: Missing NULL check

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:189]: Missing NULL chk

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:199]: NULL chk

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:265]: NULL ptr deref if `drv->builder == NULL`

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:181]: This is dangerous to do in here. When creating the driver multiple times, it would init multiple times. And every time it would change some static variables in global memory. I suggest to completely eleminate this call and leave it to downstream apps to properly init() and deinit(), like we decided to do for SDL2 driver.

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:139]: It is not clear to me why this needs to be static. Could you pls. add a comment?

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:63]: As `stride` used twice, consider making it `static const` in a global context.

**Review by derselbst — COMMENTED** [review]
I gave it a try. Unfortunately pipewire 0.3.24 fails to create the driver on my machine:

```
[E][pipewire.c:335 add_interface()] can't get Spa:Pointer:Interface:Log interface -2
[E][pipewire.c:368 load_journal_logger()] can't get log interface -2
[E][pipewire.c:335 add_interface()] can't get Spa:Pointer:Interface:CPU interface -2
[E][loop.c:85 pw_loop_new()] loop 0x60400000cc50: can't make support.system handle: No such file or directory
fluidsynth: error: Failed to allocate PipeWire loop
```

There's a similar but shorter error in 0.3.37:

```
[E][loop.c:84 pw_loop_new()] loop 0x60400000f750: can't make support.system handle: No such file or directory
fluidsynth: error: Failed to allocate PipeWire loop
```

Don't know, what it means. No time to investigate further. In any case, fluidsynth crashes afterwards, because of NULL ptr deref. Hence I flagged some issues in your code, see below. Sry for being so nit-picky.

Sadly, only by including the pipewire header we get  a bunch of `warning: cast discards ‘const’ qualifier from pointer target type`

File: src/drivers/fluid_pipewire.c
Comment (SubhadeepJasu) [reply src/drivers/fluid_pipewire.c:181]: How will we access the init and deinit functions downstream?

Comment (derselbst) [reply src/drivers/fluid_pipewire.c:181]: You'll have to include the header.

File: src/drivers/fluid_pipewire.c
Comment (sykhro) [reply src/drivers/fluid_pipewire.c:139]: it doesn't, actually, let me fix this

File: src/drivers/fluid_pipewire.c
Comment (sykhro) [reply src/drivers/fluid_pipewire.c]: Done, will need to find a way to detect that pw hasn't been init.

**Comment by sykhro** [comment]
> I gave it a try. Unfortunately pipewire 0.3.24 fails to create the driver on my machine:
> 
> ```
> [E][pipewire.c:335 add_interface()] can't get Spa:Pointer:Interface:Log interface -2
> [E][pipewire.c:368 load_journal_logger()] can't get log interface -2
> [E][pipewire.c:335 add_interface()] can't get Spa:Pointer:Interface:CPU interface -2
> [E][loop.c:85 pw_loop_new()] loop 0x60400000cc50: can't make support.system handle: No such file or directory
> fluidsynth: error: Failed to allocate PipeWire loop
> ```
> 
> There's a similar but shorter error in 0.3.37:
> 
> ```
> [E][loop.c:84 pw_loop_new()] loop 0x60400000f750: can't make support.system handle: No such file or directory
> fluidsynth: error: Failed to allocate PipeWire loop
> ```
> 
> Don't know, what it means. No time to investigate further. In any case, fluidsynth crashes afterwards, because of NULL ptr deref. Hence I flagged some issues in your code, see below. Sry for being so nit-picky.
> 
> Sadly, only by including the pipewire header we get a bunch of `warning: cast discards ‘const’ qualifier from pointer target type`

Is PipeWire your audio server?

**Comment by derselbst** [comment]
> Is PipeWire your audio server?

I don't think so. I haven't made any special setup. Only installed pipewire-alsa and hoped it would work out-of-the-box. Perhaps I'll find time to try pipewire-jack later on.

**Comment by sykhro** [comment]
> > Is PipeWire your audio server?
> 
> I don't think so. I haven't made any special setup. Only installed pipewire-alsa and hoped it would work out-of-the-box. Perhaps I'll find time to try pipewire-jack later on.

I see. `pipewire` is the main library, which needs to be set up as the system audio server. The other packages act as some sort of translation layer that make things work oob: `pipewire-alsa` is for ALSA apps, `pipewire-pulse` is for pulseaudio apps, `pipewire-jack` is for JACK apps.

**Comment by SubhadeepJasu** [comment]
Yeah you need the one that just says "pipewire". The others simply act as "adapters".

File: src/drivers/fluid_pipewire.c
Comment (SubhadeepJasu) [reply src/drivers/fluid_pipewire.c]: oh

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:167]: `fluid_settings_dupstr` requires you to `FLUID_FREE` the strings later on, which you currently not do.

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:177]: You're leaking the `drv` object here. Consider using `goto driver_cleanup`.

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:199]: Consider `goto driver_cleanup`

File: src/drivers/fluid_pipewire.c
Comment (derselbst) [inline src/drivers/fluid_pipewire.c:296]: Those new `audio.pipewire.*`  have no options registered (what we usually do with `fluid_settings_add_option()`). Is that intended? If they only ever default to the standard option, do we actually need them?
In case we do need them, they should be documented in /doc/fluidsettings.xml. Just placing a link to the pipewire documentation should be sufficient.

**Review by derselbst — COMMENTED** [review]
There hasn't been any activity for quite some time, so it seems to me that this is ready for merge. I gave it one more review and found a few more tiny issues, pls. see below. 

Also, I updated pipewire on my machine to 0.3.40 and now at least the NULL deref and crash is gone. I still haven't figured out how to get it running. Setting up a VM for this is a bit of an overkill, sry. But if it works well for you I would be happy to take this.

File: src/drivers/fluid_pipewire.c
Comment (sykhro) [reply src/drivers/fluid_pipewire.c:304]: Yeah, I didn't set options because application developers can use different values here. e.g. when using fluidsynth as the audio engine in a game, one can set "audio.pipewire.media-role" to "Game" as per [the freedesktop specification](https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/Developer/Clients/ApplicationProperties/). This is for integration with the system's audio mixer. Not sure what to do here, besides copying all the values from the spec..

Comment (derselbst) [reply src/drivers/fluid_pipewire.c:304]: NP, I'll take care of the documentation after merge.

**Review by derselbst — APPROVED** [review]
LGTM. I'll merge it now. This will be part of fluidsynth 2.3.0. Not sure when it'll be released, we have a few other open issues.
Thanks for contributing!

**Comment by shmerl** [comment]
So should pipewire be an option for the fluidsynth now?

I see this with fluidsynth 2.3.1 (in Debian testing):

```
fluidsynth -a help
-a options (audio driver):
   'alsa','file','jack','oss','pulseaudio','sdl2'
```

**Pull request merged**
